[CPU][ArmSME] Add (initial) tiling and lowering pipeline for ArmSME

MacDue opened the pull request:
This adds initial support for ArmSME. Currently, this only targets f32 and f64 matmuls (as there's little support for anything else in currently available tools). When `+sme` is set f64 and f32 matmuls will be tiled (at the kernel dispatch level), into sizes that match the SME virtual tiles. The lowerings after that are all standard, until the final LLVM lowerings, where the ArmSME passes are enabled to allow lowering the 2D scalable vector types to ArmSME intrinsics.

Example compiler usage:

```sh
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-llvmcpu-target-cpu-features=+sve,+sme \
    matmul.mlir -o $output_file
```

MacDue committed:
[CPU][ArmSME] Add tiling and lowering pipeline for ArmSME

This adds initial support for ArmSME. Currently, this only targets f32
and f64 matmuls (as there's little support for anything else in
currently available tools). When `+sme` is set f64 and f32 matmuls will
be tiled (at the kernel dispatch level), into sizes that match the SME
virtual tiles. The lowerings after that are all standard, until the
final LLVM lowerings, where the ArmSME passes are enabled to allow
lowering the 2D scalable vector types to ArmSME intrinsics.

ArmSME also requires some SME compiler-rt builtins, which need to be
provided by setting `IREE_ARM_SME_COMPILER_RT_BUILTINS_STATIC_LIB` to
a path that contains them. Though at the moment stubs can also be used
instead of the actual builtins.

Example compiler usage:

```sh
export IREE_ARM_SME_COMPILER_RT_BUILTINS_STATIC_LIB="<arm_sme_builtins.o>"
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-llvmcpu-target-cpu-features=+sve,+sme \
    matmul.mlir -o $output_file
```

c-rhodes commented on compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:
```diff
@@ -913,6 +914,36 @@ class ExpandMulSIExtended : public OpRewritePattern<arith::MulSIExtendedOp> {
+        llvmFuncOp->getParentOfType<IREE::HAL::ExecutableVariantOp>();
+
+    if (variantOp.getObjectsAttr())
+      return failure();
```
could the objects attr contain something other than the SME compiler-rt lib? Just wondering if this check should be tighter.

MacDue replied:
I admit this check is a little lazy, it probably should specifically look for the `compiler-rt` object file :)
 I think it's pretty unlikely that something other than `compiler-rt` would be linked though, as I think this is typically used for micro-kernels.

MacDue replied:
I'll fix this up soon :+1:

c-rhodes reviewed (commented):
Thanks for the patch Ben, this is exciting! It's taken a lot of work to get here, mostly in core MLIR, but good to see the integration doesn't require significant changes on the IREE side. Mostly LGTM but I'll take another look once the deps land, cheers

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -949,6 +955,40 @@ static void getDefaultMatmulVectorSizes(
   return;
 }
 
+static FailureOr<Type> nonWideningMatmulElementType(linalg::LinalgOp op) {
```
doc

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -949,6 +955,40 @@ static void getDefaultMatmulVectorSizes(
+
+  // TODO(macdue): Other element types (there is little support for anything
+  // other than f32 and f64 yet).
+}
```
@pzread, another example of an ad-hoc TSS strategy :)

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1005,7 +1053,12 @@ static SizesAndScalableFlags getMatmulVectorSizes(func::FuncOp entryPointFn,
+      // sizes must match SME tiles (and cannot be arbitrarily resized).
+      if (numScalableDims >= 2 && scalableTileFlags[i]) {
+        continue;
+      }
```
Yeah, this after-the-fact tile size adjustments are a pain. We are trying to get rid of those with a new tile size selection infra.

hanhanW replied:
I thought the "infra" is more like providing you a way to configure tile sizes. The adjustment is quite independent to the infra. The size adjustment is more about how you compute tile sizes. You can always get the tile sizes you set without adjustment, just don't call those functions?

Anyway, I agree that it is a pain; I've been working progress on removing historical heuristics. I just wanna flag that you will run into the same situation even we have a new tile size selection infra. You will either call the existing methods or re-implement same methods in a different way. The new tile size selection infra does not address the adjustment issue to me. It is an infra, but not a "fix".

pzread replied:
I think one point we can improve is not having these target-dependent decisions in the generic utilities like `getMatmulVectorSizes` here (and [this](https://github.com/openxla/iree/blob/2529fb37c76c89c0306a4bb45f8bdcdf8baeaf16/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1057-L1062)). This can be one reason confusing as we are trying to have a generic tile size selection workflow for each op, but target-specific logic/parameters/heuristics are segregated in each step (distribution, cache, vector tile selection steps)

I'm trying something like dispatching the tile size selection workflow based on the target architecture, then each workflow only needs to take care about its target-specific parameters/logic/heuristics. Then we can factor out common code into target-independent utilities so each workflow can be implemented concisely. Nothing concretes yet but just an idea.

dcaballe reviewed (approved):
Great milestone! LGTM but please wait for Scott or someone else to review the compile_rt dep and the benchmark side of things

ScottTodd commented on compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:
```diff
@@ -913,6 +914,36 @@ class ExpandMulSIExtended : public OpRewritePattern<arith::MulSIExtendedOp> {
+
+    if (!sme_compiler_rt_lib)
+      return llvmFuncOp.emitError(
+          "IREE_ARM_SME_COMPILER_RT_BUILTINS_STATIC_LIB must be set!");
```
> ArmSME also requires some SME compiler-rt builtins, which need to be provided by setting `IREE_ARM_SME_COMPILER_RT_BUILTINS_STATIC_LIB` to a path that contains them. Though at the moment stubs can also be used instead of the actual builtins.

Can you elaborate on what is included in this static lib? Are the sources for that redistributable? Could we include it as part of our build process, rather than rely on an environment variable? We have similar dependencies for some CPU intrinsics (musl) and GPU SDKs / libraries (CUDA, HIP/ROCm, WebGPU with Tint, etc.).

cc @benvanik , @stellaraccident

MacDue replied:
It's the Arm SME ABI support routines, added to compiler-rt with this commit: https://github.com/llvm/llvm-project/commit/f445be9790f9998e9048fc2a12402f3e213c079f. If a function (or dispatch) is marked as using 'ZA' (the SME accumulator state), the LLVM backend plants calls to ABI support routines to ensure this state is preserved between calls.

ScottTodd replied:
Maybe we could build compiler-rt (when ARM is in `LLVM_TARGETS_TO_BUILD`, see https://github.com/openxla/iree/blob/main/build_tools/cmake/iree_llvm.cmake).

MacDue replied:
I feel it may be a little too early to setup a proper build for `compiler-rt` just yet. We're not actually making use of the functionality it provides yet, so it's simpler to just link in a stub object file. 

Currently, building the actual `compiler-rt` routines with SME requires a pretty recent compiler (clang-16, I believe), otherwise it'll just exclude them from the build. They also depend on either glibc or newlib (depending on if it's a bare-metal build or not), which may complicate things a little more. 

I definitely think we should set this up properly at some point, but I feel it may be simpler to leave this while things are still in flux?

ScottTodd replied:
> We're not actually making use of the functionality it provides yet, so it's simpler to just link in a stub object file.

Could the `return llvmFuncOp.emitError` be downgraded to a warning then? I'm worried that compiling for one CPU architecture will work out of the box but compiling for another will throw an error (and the error would be partway through compilation, not at startup time).

I'd also like to hear from @benvanik or @stellaraccident

ScottTodd replied:
Oh, and if this is left open, can you file an issue to track "setting it up properly"?

MacDue replied:
> Could the return llvmFuncOp.emitError be downgraded to a warning then? I'm worried that compiling for one CPU architecture will work out of the box but compiling for another will throw an error (and the error would be partway through compilation, not at startup time).

It still would throw an error (just during the linking pass).

MacDue committed:
Address review comments and solve merge conflicts

- Add doc comments
- Explicitly check for the SME routines in the objects attribute
- Also, ensure when tiling the linalg op that it's one of the ops known
  to be supported by ArmSME

MacDue committed:
Apply bazel linter changes

benvanik reviewed (changes requested):
The compiler-rt dep is a hard blocker for this in IREE - we cannot have additional deps like this and what's in it seems to indicate a design issue (not major I think, but at least needs to be solved before landing) if things like abort() and platform-specific (?) register save/restore are being inserted during code generation. IREE executables are portable across all platforms and are never allowed to abort() or make any kind of syscalls or globally impacting operations.

Ideally upstream shouldn't be inserting aborts in user programs, or should be inserting them in a way we can remove them here by turning them into non-zero returns from the dispatch functions (gracefully failing instead of bringing down the entire user process). I suspect this isn't too hard to fix but may require a pass/pattern to do that substitution if upstream is really inserting aborts (find your aborts and replace with return 1 may be enough). That pass could live in IREE if needed but we'd want to have a stable thing to match on emitted by the upstream code.

How platform-specific is the register save/restore code? If it's always the same on all operating systems (and the platform-specific comments in the compiler-rt impl are just about the abort) then it may just mean moving that out of compiler-rt and into LLVM IR you emit. We do have a mechanism for builtins (https://github.com/openxla/iree/blob/main/runtime/src/iree/builtins/device/README.md) that we link into codegen binaries where the save/restore could be moved so long as they are portable, but IMO not requiring any deps or users to do such things for what amounts to a dozen assembly instructions would be ideal and help with hermetic deployments. This again may be a pass we carry in IREE that just adds those two functions in LLVM IR with inline asm blocks only when one or more dispatches uses SME.

Happy to chat more about the exact details, but the TLDR is no abort() (must return non-zero if failures are unrecoverable) and no compiler-rt dep for save/restore. There are ways around both and you may have some approaches you've discussed already. It's a blocker here but I don't think a large lift to have a pass that does the right thing, and we're happy to carry it if needed. Happy to discuss more on the iree-discuss list, discord, or the next community meeting!

MacDue committed:
Add ArmSME ABI stubs to `libdevice`

Note that use use of these stubs is safe for the code IREE currently
generates, as each dispatch region is self-contained. SME dispatches do
not call out to other SME-enabled functions, so ZA-state does not need
to be preserved.

The LLVM backend still plants calls to these functions, so these
definitions are simply to satisfy the linker, but these functions won't
ever be called.

Preserving ZA state between OS threads is handled by the kernel, and
does not depend on these ABI routines.
See: https://docs.kernel.org/arch/arm64/sme.html

c-rhodes commented on runtime/src/iree/builtins/device/BUILD.bazel:
```diff
@@ -51,6 +51,13 @@ endif()
 # TODO(benvanik): rule for building a matrix of bitcode files.
 
+iree_bitcode_library(
+    name = "libdevice_device_aarch64_sme",
```
`libdevice_aarch64_sme` for consistency?

MacDue replied:
Whoops, that was a typo! I was blind to the double 'device' there :sweat_smile: 
I'll fix that up 👍

c-rhodes commented on runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c:
```diff
@@ -0,0 +1,45 @@
+typedef struct TPIDR2_block TPIDR2_block_t;
+
+IREE_DEVICE_EXPORT sme_state_t __arm_sme_state(void) {
+  // TODO(macdue): Implement.
```
Given we don't expect SME-enabled dispatch functions with live ZA state to call other such functions, I think we can remove the TODO's, as these stubs are sufficient?

MacDue replied:
I think it's still a TODO, even if we don't need these functions (yet), that might not always be the case.

benvanik replied:
with TODOs we try to make them actionable (in case the originating author disappears) - a comment of "// No-op today as not needed." would be informative/durable and let people know this file is good as-is, while the TODO without context makes it look like it's not done yet.

MacDue committed:
Fix typo `libdevice_device` -> `libdevice`

c-rhodes reviewed (approved):
LGTM cheers

MacDue committed:
Update tests to allow for dropping unit dims in vector transfers

MacDue committed:
Fix bazel build

benvanik reviewed (approved):
Nice! Thanks for the changes. We can evaluate in the future if we move these into the compiler to emit instead of needing a unique bitcode file but this at least puts things in the control of the compiler and will be easy to fix up in the future. One nit on the TODOs and otherwise LGTM!

hanhanW commented on tests/e2e/matmul/BUILD.bazel:
```diff
@@ -62,6 +62,31 @@ py_binary(
 
+# LLVMCPU, non-data-tiling, no microkernels, ArmSME
+[iree_generated_trace_runner_test(
+    name = "e2e_matmul_arm_sme_nondt_%s_%s" % (dtype, size),
```
I think it is okay to add a new entry for the test suite. But can we remove the original one? E.g., can we remove line 50 in this file?

MacDue replied:
The tests added here are just for `SME`, line 50 adds tests for `SVE`. I put the `SME` tests in their own section as (for now), SME is a bit more of a special case :)

hanhanW replied:
oops, I misread it. Sorry about that and good catch!

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -962,16 +967,67 @@ static void getDefaultMatmulVectorSizes(
+  // Double check the operation is one that is supported for lowering to ArmSME.
+  if (!llvm::isa<linalg::MatmulOp, linalg::MatmulTransposeAOp, linalg::FillOp>(
+          op))
+    return;
```
Do we need to check `linalg:FillOp`? It is not common to have a single fill op in dispatch. I thought this function is mostly used for matmuls?

MacDue replied:
Nope, good point, this only applies to `LinalgContractionOps`, so won't ever see a fill :+1:

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -962,16 +967,67 @@ static void getDefaultMatmulVectorSizes(
+       llvm::concat<Value>(op.getRegionInputArgs(), op.getRegionOutputArgs())) {
+    inputAndOutputElementTypes.push_back(v.getType());
+  }
+  if (inputAndOutputElementTypes.empty() ||
```
Can we remove the `.empty()` check? I think this is an invalid linalg op if it happens.

MacDue replied:
I've changed this to a debug assert :+1:

hanhanW reviewed (approved):
Awesome work!

MacDue committed:
Review fixups
